core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java

```diff
@@ -0,0 +1,186 @@
+            assertThat("suggestion builder copy's hashcode is different from original hashcode", firstBuilder.hashCode(), equalTo(thirdBuilder.hashCode()));
+            assertTrue("equals is not symmetric", thirdBuilder.equals(secondBuilder));
+            assertTrue("equals is not symmetric", thirdBuilder.equals(firstBuilder));
+        }
+    }
+
+    private SB mutate(SB firstBuilder) throws IOException {
+        SB mutation = serializedCopy(firstBuilder);
+        assertNotSame(mutation, firstBuilder);
+        if (randomBoolean()) {
+            // change one of the common SuggestionBuilder parameters
+            switch (randomIntBetween(0, 6)) {
+            case 0:
+                mutation.text(randomValueOtherThan(mutation.text(), () -> randomAsciiOfLengthBetween(2, 20)));
+                break;
+            case 1:
+                mutation.prefix(randomValueOtherThan(mutation.prefix(), () -> randomAsciiOfLengthBetween(2, 20)));
+                break;
+            case 2:
+                mutation.regex(randomValueOtherThan(mutation.regex(), () -> randomAsciiOfLengthBetween(2, 20)));
+                break;
+            case 3:
+                mutation.field(randomValueOtherThan(mutation.field(), () -> randomAsciiOfLengthBetween(2, 20)));
+                break;
+            case 4:
+                mutation.analyzer(randomValueOtherThan(mutation.analyzer(), () -> randomAsciiOfLengthBetween(2, 20)));
+                break;
+            case 5:
+                mutation.size(randomValueOtherThan(mutation.size(), () -> randomIntBetween(1, 20)));
+                break;
+            case 6:
+                mutation.shardSize(randomValueOtherThan(mutation.shardSize(), () -> randomIntBetween(1, 20)));
+                break;
+            }
+        } else {
+            mutateSpecificParameters(firstBuilder);
+        }
+        return mutation;
+    }
+
+    /**
+     * take and input {@link SuggestBuilder} and return another one that is different in one aspect (to test non-equality)
+     */
+    protected abstract void mutateSpecificParameters(SB firstBuilder) throws IOException;
+
+    @SuppressWarnings("unchecked")
+    protected SB serializedCopy(SB original) throws IOException {
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.writeSuggestion(original);;
+            try (StreamInput in = new NamedWriteableAwareStreamInput(StreamInput.wrap(output.bytes()), namedWriteableRegistry)) {
+                return (SB) in.readSuggestion();
+            }
+        }
+    }
+
+    protected static <T> void maybeSet(Consumer<T> consumer, T value) {
+        if (randomBoolean()) {
+            consumer.accept(value);
+        }
+    }
+
+    /**
+     * helper to get a random value in a certain range that's different from the input
+     */
+    protected static <T> T randomValueOtherThan(T input, Supplier<T> randomSupplier) {
+        T randomValue = null;
```

Comment:
I would have loved to have something like this for the *Sort tests and came up with a less elegant solution. Can we have that in some more generic helper class?